src/backend/distributed/commands/utility_hook.c

```diff
@@ -912,6 +912,22 @@ ExecuteDistributedDDLJob(DDLJob *ddlJob)
 		{
 			CommitTransactionCommand();
 			StartTransactionCommand();
+
+			/*
+			 * Even if we started a new transaction, now we still hold
+			 * ExclusiveLock (acquired by VirtualXactLockTableInsert) on new
+			 * vxid and it is not released yet.
+			 *
+			 * Moreover, remote connections to localhost (that are opened by
+			 * Citus for shard level index commands) might enter into dead-lock
+			 * since they would attempt to acquire ShareLock on the same vxid
+			 * since postgres calls WaitForOlderSnapshots function when executing
+			 * CREATE INDEX (or REINDEX) CONCURRENTLY commands.
+			 *
+			 * To prevent self-deadlock, we release lock on vxid by explicitly
+			 * calling VirtualXactLockTableCleanup.
+			 */
+			VirtualXactLockTableCleanup();
```

Comment:
It's kind of odd to me that this happens *after* StartTransactionCommand. Is there a reason for that?

Reply:
yeah, it's because now we hold lock on vxid for new xact, so committing & starting the xact doesn't have any effect on this particular deadlock around vxid

Reply:
for example,

// vxid was 7/40
CommitXact();
StartXact();
// now vxid is 7/41 but remote connections would also have 7/41 since that new xact is not committed yet

Reply:
This seems extremely hacky. The right answer might be to not start a new transaction at all, or to figure out what's going on with the create extension logic.

Reply:
I wonder if we should fall-back to using local execution here. I know it was crashing when we do but which one is easier to fix (solving the deadlock or crash with local execution)

Reply:
>or to figure out what's going on with the create extension logic.

Will first look into that